src/dynamic_boundary_conditions/tide_data.py

```diff
@@ -1,148 +1,458 @@
+        tide_df.insert(loc=2, column='longitude', value=query_param['long'])
+        return tide_df
+
+
+async def get_tide_data_for_requested_period(
+        query_param_list: List[Dict[str, Union[str, int]]],
+        url: str = 'https://api.niwa.co.nz/tides/data'):
+    """
+    Iterate over the list of API query parameters to fetch high and low tide data for the requested period.
+
+    Parameters
+    ----------
+    query_param_list : List[Dict[str, Union[str, int]]]
+        A list of api query parameters used to retrieve high and low tide data for the entire requested period.
+    url: str = 'https://api.niwa.co.nz/tides/data'
+        Tide API HTTP request url.
+    """
+    tasks = []
     async with aiohttp.ClientSession() as session:
-        tasks = []
-        for params in params_list:
-            tasks.append(get(session=session, url=url, params=params))
-        # asyncio.gather() will wait on the entire task set to be completed.
-        data_dict = await asyncio.gather(*tasks, return_exceptions=True)
-        data_list = [_dict['values'] for _dict in data_dict]
-        df = pd.DataFrame([_item for _list in data_list for _item in _list])
-        return df
-
-
-def get_highest_tide_dataframe(df: pd.DataFrame, before_n_days: int, after_n_days: int) -> pd.DataFrame:
-    """ takes dataframe and finds data either side of peak is required (in days) """
-    # sort dataframe by tide value.
-    df['time'] = pd.to_datetime(df['time'])
-    df['time'] = df['time'].dt.tz_convert(tz='Pacific/Auckland')
-    df = df.sort_values(by=['value'], ascending=False).reset_index(drop=True)
-    df.set_index('time', inplace=True)
-    print("Highest tide date is: {}".format(df.index[0]))
-    # setting upper and lower index, index 0 is the highest tide date
-    lower_boundary = (df.index[0] - timedelta(days=max(before_n_days, 0)))
-    upper_boundary = (df.index[0] + timedelta(days=max(after_n_days, 0)))
-    # protect index boundary.
-    assert lower_boundary.strftime('%Y-%m-%d') in df.index, \
-        f"Index overflow: lower boundary {lower_boundary} is not in dataframe index."
-    assert upper_boundary.strftime('%Y-%m-%d') in df.index, \
-        f"Index overflow: upper boundary {upper_boundary} is not in dataframe index."
-    # get dataframe between lower and upper boundary date.
-    df = (df.sort_values(by=['time'])
-            .loc[lower_boundary.strftime('%Y-%m-%d'):upper_boundary.strftime('%Y-%m-%d')])
-    # convert dataframe schema for sql
-    df.reset_index(inplace=True)
-    df['time'] = df['time'].astype(str)
-    df.columns = ['Datetime_(NZST)', 'Tide_value_(m)']
-    return df
-
-
-def save_to_database(connect, df: pd.DataFrame, table_name: str, if_exists: str = 'replace'):
-    """ save dataframe to database """
-    df.to_sql(table_name, connect, index=False, if_exists=if_exists)
+        # Create a list of tasks that fetch tide data for each query parameter
```

Comment:
We may be using a different method of asynchronous tasks soon but keep this as is for now, no reason to change until this is integrated with Flask and Celery, and maybe not even then.

Reply:
Ok.